// html-reporter/build/lib/constants/database.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DB_COLUMN_INDEXES = exports.DATABASE_URLS_JSON_NAME = exports.LOCAL_DATABASE_NAME = exports.DB_SUITES_TABLE_NAME = exports.DB_MAX_AVAILABLE_PAGE_SIZE = exports.SUITES_TABLE_COLUMNS = exports.DB_COLUMNS = exports.DB_TYPES = void 0;
// TODO: change to enums
exports.DB_TYPES = { int: 'INT', text: 'TEXT' };
exports.DB_COLUMNS = {
    SUITE_PATH: 'suitePath',
    SUITE_NAME: 'suiteName',
    NAME: 'name',
    SUITE_URL: 'suiteUrl',
    META_INFO: 'metaInfo',
    HISTORY: 'history',
    DESCRIPTION: 'description',
    ERROR: 'error',
    SKIP_REASON: 'skipReason',
    IMAGES_INFO: 'imagesInfo',
    SCREENSHOT: 'screenshot',
    MULTIPLE_TABS: 'multipleTabs',
    STATUS: 'status',
    TIMESTAMP: 'timestamp',
    DURATION: 'duration',
    ATTACHMENTS: 'attachments'
};
exports.SUITES_TABLE_COLUMNS = [
    { name: exports.DB_COLUMNS.SUITE_PATH, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.SUITE_NAME, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.NAME, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.SUITE_URL, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.META_INFO, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.HISTORY, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.DESCRIPTION, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.ERROR, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.SKIP_REASON, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.IMAGES_INFO, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.SCREENSHOT, type: exports.DB_TYPES.int },
    { name: exports.DB_COLUMNS.MULTIPLE_TABS, type: exports.DB_TYPES.int },
    { name: exports.DB_COLUMNS.STATUS, type: exports.DB_TYPES.text },
    { name: exports.DB_COLUMNS.TIMESTAMP, type: exports.DB_TYPES.int },
    { name: exports.DB_COLUMNS.DURATION, type: exports.DB_TYPES.int },
    { name: exports.DB_COLUMNS.ATTACHMENTS, type: exports.DB_TYPES.text }
];
exports.DB_MAX_AVAILABLE_PAGE_SIZE = 65536; // helps to speed up queries
exports.DB_SUITES_TABLE_NAME = 'suites';
exports.LOCAL_DATABASE_NAME = 'sqlite.db';
exports.DATABASE_URLS_JSON_NAME = 'databaseUrls.json';
exports.DB_COLUMN_INDEXES = exports.SUITES_TABLE_COLUMNS.reduce((acc, { name }, index) => {
    acc[name] = index;
    return acc;
}, {});
