// html-reporter/build/lib/constants/defaults.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configDefaults = exports.CIRCLE_RADIUS = void 0;
const diff_modes_1 = require("./diff-modes");
const view_modes_1 = require("./view-modes");
const save_formats_1 = require("./save-formats");
exports.CIRCLE_RADIUS = 150;
exports.configDefaults = {
    baseHost: '',
    commandsWithShortHistory: [],
    customGui: {},
    customScripts: [],
    defaultView: view_modes_1.ViewMode.ALL,
    diffMode: diff_modes_1.DiffModes.THREE_UP.id,
    enabled: false,
    errorPatterns: [],
    lazyLoadOffset: null,
    metaInfoBaseUrls: {},
    path: '',
    plugins: [],
    pluginsEnabled: false,
    saveErrorDetails: false,
    saveFormat: save_formats_1.SaveFormat.SQLITE,
    yandexMetrika: {
        counterNumber: 99267510
    },
    staticImageAccepter: {
        enabled: false,
        repositoryUrl: '',
        pullRequestUrl: '',
        serviceUrl: '',
        meta: {},
        axiosRequestOptions: {}
    }
};
