// html-reporter/build/lib/constants/test-statuses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.COMMITED = exports.STAGED = exports.FAIL_ERROR = exports.UPDATED = exports.SKIPPED = exports.ERROR = exports.FAIL = exports.SUCCESS = exports.RUNNING = exports.QUEUED = exports.IDLE = exports.TestStatus = void 0;
var TestStatus;
(function (TestStatus) {
    TestStatus["IDLE"] = "idle";
    TestStatus["QUEUED"] = "queued";
    TestStatus["RUNNING"] = "running";
    TestStatus["SUCCESS"] = "success";
    TestStatus["FAIL"] = "fail";
    TestStatus["ERROR"] = "error";
    TestStatus["SKIPPED"] = "skipped";
    TestStatus["UPDATED"] = "updated";
    /**
     * @note used by staticImageAccepter only
     */
    TestStatus["STAGED"] = "staged";
    /**
     * @note used by staticImageAccepter only
     */
    TestStatus["COMMITED"] = "commited";
    /**
     * @note used in new UI only for rendering icons
     */
    TestStatus["RETRY"] = "retry";
    /** @note used to display tests that have both failed screenshots and errors */
    TestStatus["FAIL_ERROR"] = "fail_error";
})(TestStatus || (exports.TestStatus = TestStatus = {}));
exports.IDLE = TestStatus.IDLE;
exports.QUEUED = TestStatus.QUEUED;
exports.RUNNING = TestStatus.RUNNING;
exports.SUCCESS = TestStatus.SUCCESS;
exports.FAIL = TestStatus.FAIL;
exports.ERROR = TestStatus.ERROR;
exports.SKIPPED = TestStatus.SKIPPED;
exports.UPDATED = TestStatus.UPDATED;
exports.FAIL_ERROR = TestStatus.FAIL_ERROR;
/**
 * @note used by staticImageAccepter only
 */
exports.STAGED = TestStatus.STAGED;
/**
 * @note used by staticImageAccepter only
 */
exports.COMMITED = TestStatus.COMMITED;
