// html-reporter/build/lib/gui/app.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.App = void 0;
const tool_runner_1 = require("./tool-runner");
class App {
    static create(args) {
        return new this(args);
    }
    constructor(args) {
        this._toolRunner = tool_runner_1.ToolRunner.create(args);
    }
    get data() {
        return this._toolRunner.tree;
    }
    async initialize() {
        return await this._toolRunner.initialize();
    }
    async finalize() {
        return this._toolRunner.finalize();
    }
    async run(tests) {
        return this._toolRunner.run(tests);
    }
    getTestsDataToUpdateRefs(imageIds = []) {
        return this._toolRunner.getTestsDataToUpdateRefs(imageIds);
    }
    getImageDataToFindEqualDiffs(imageIds = []) {
        return this._toolRunner.getImageDataToFindEqualDiffs(imageIds);
    }
    async updateReferenceImage(failedTests = []) {
        return this._toolRunner.updateReferenceImage(failedTests);
    }
    async undoAcceptImages(imageIds) {
        return this._toolRunner.undoAcceptImages(imageIds);
    }
    async findEqualDiffs(data) {
        return this._toolRunner.findEqualDiffs(data);
    }
    addClient(connection) {
        this._toolRunner.addClient(connection);
    }
}
exports.App = App;
